### ISSUE TITLE
lldpd in EPEL and Fedora repositories

### ISSUE THREAD

**Issue by hogarthj** [vincentbernat/lldpd#227]
Hi,

This is a courtesy ticket as a friendly hello and heads up :)

I've just put this through package review in Fedora and will be maintaining it in the repositories there for Fedora and EPEL.

https://bugzilla.redhat.com/show_bug.cgi?id=1438853

If you like I'll let you know when it's in the stable repos so that you can include the update in your install instructions.

This does build for EPEL6 as well with only a few minor changes in a patch to handle the older bash and autoconf there.

```
diff --git a/configure.ac b/configure.ac
index 013eca0..a5af473 100644
--- a/configure.ac
+++ b/configure.ac
@@ -5,10 +5,10 @@
 ### Base configuration
 
 # Configure autoconf
-AC_PREREQ([2.64])
+AC_PREREQ([2.63])
 
 AC_INIT([lldpd],
-        [m4_esyscmd_s([./get-version])],
+        [2017-04-05],
         [https://github.com/vincentbernat/lldpd/issues],
         [lldpd],
         [http://vincentbernat.github.io/lldpd/])
@@ -31,9 +31,18 @@ AM_INIT_AUTOMAKE([foreign subdir-objects -Wall -Werror tar-ustar])
 AM_MAINTAINER_MODE
 m4_ifdef([AM_SILENT_RULES], [AM_SILENT_RULES(yes)])
 
-# Automake 1.12...
-m4_pattern_allow([AM_PROG_AR])
-AM_PROG_AR
+
+dnl Autoconf 2.64 introduced AS_VAR_COPY, so introduce a compat version
+dnl for Autoconf 2.63.
+m4_ifndef([AS_VAR_COPY],
+  [m4_define([AS_VAR_COPY],
+    [AS_LITERAL_IF([$1[]$2], [$1=$$2], [eval $1=\$$2])])])
+
+dnl Autoconf 2.64 introduced AS_VAR_APPEND, so introduce a compat version
+dnl for Autoconf 2.63.
+m4_ifndef([AS_VAR_APPEND],
+  [m4_define([AS_VAR_APPEND],
+      [$1=$$1$2])])
 
 # Configure libtool
 LT_INIT
@@ -193,14 +202,7 @@ PKG_CHECK_MODULES([libbsd], [libbsd-overlay], [
 AC_REPLACE_FUNCS([setproctitle])
 AC_CHECK_FUNCS([setproctitle_init])
 # Other functions
-AC_REPLACE_FUNCS([strlcpy
-                  strnlen
-                  strndup
-                  strtonum
-                  getline
-                  asprintf
-                  vsyslog
-                  daemon])
+AC_REPLACE_FUNCS([strlcpy strnlen strndup strtonum getline asprintf vsyslog daemon])
 # Optional functions
 AC_CHECK_FUNCS([setresuid setresgid])
```

I'm happy to maintain it as a maintainer patch downstream so that EPEL6 gets built, or to work with you for something that can build the autoconf 2.63 shipped with EL6 still (hey at least EL5 is EOL now!).

Oh and embedded libevent is not needed on EL6 anymore as Red Hat included libevent2 as of EL6.8 which this builds cleanly against.

Thanks for developing this tool :)

James

**Comment by vincentbernat** [comment]
Great!

As for autoconf/automake, you don't have to rebuild them (but maybe, you have a policy to systematically rebuild autoconf).

I'll have a look at other issues later since I am currently a bit scarce in time. Let's keep the ticket open. I'll update the documentation next week.

**Comment by hogarthj** [comment]
Fedora policies are to use the tarball from releases unless there's a legal issue or something like that or it really is impossible, so it's better for us/me if I have a small patch and have the same autoconf/configure/make behaviour on all branches between fedora and epel to keep consistent behaviour.

No hurries on things - I know well how time can run away ... this was more a friendly face than anything else.

**Comment by vincentbernat** [comment]
About the tarball, if you use https://media.luffy.cx/files/lldpd/lldpd-0.9.7.tar.gz, you'll get a version with autotools stuff already generated. This is the "official" tarball (also linked from the website). You get a GPG signature if you append `.gpg` to it.

Also, an easy way to test lldpd is to create a veth device: `ip link add type veth && ip link set up dev veth0 && ip link set up dev veth1`.

For `chroot()`/`chdir()`, I don't understand the warning since `chdir()` is done just after `chroot()` (in `src/daemon/priv.c`).

**Comment by hogarthj** [comment]
That is such a flakey test in rpmlint which frequently pulls false positives (there's a longstanding open issue for it) that it can safely be ignored.

Thanks for the pointer about that tarball, might well switch over to that then.

I'm not worried about the automated testing, it's one of the downsides of the isolated mock environment ... device tests become troublesome due to inability to mess with devices... for obvious reasons Fedora koji is pretty locked down.

You can see my spec and any patches I'm applying here:

https://src.fedoraproject.org/cgit/rpms/lldpd.git/

If there's any one who wanders over here with a bug that looks distro specific it's best to point them to this location to file with us:

https://apps.fedoraproject.org/packages/lldpd/bugs

Or if there is something you'd like me to take a look at just ping me in the issue, it's what the certbot guys often do ;)

The initial commit is Fedora/EPEL7 only but I'll be adding the EPEL6 conditionals to the spec shortly.

**Comment by vincentbernat** [comment]
The tarball change should also enable you to keep the `m4_easycmd_s` command. It's relying on a file generated when making the tarball. Otherwise, a git checkout is needed to get the version. I see a bug when not using git and not having a release tarball. I'll push a commit for that shortly.

**Comment by hogarthj** [comment]
So I took a look at this before bed.

The release tarball doesn't have the redhat directory with the example sysconfig and init file that I was copying over, but that's not much of a big deal as I can just include them separate next to the tarball.

The good news is with the libevent2-devel that they added in 6.8 it builds cleanly and no need for any patches against the release tarball at all :)

The lack of tests in the release tarball isn't an issue either since I can't run them as part of the mock build in koji anyway.

The packages are built and heading to the testing repos now.

It should be stable in Fedora in about a week, and in EPEL6 and EPEL7 in about a fortnight.

**Comment by vincentbernat** [comment]
For Fedora/EPEL 7, the systemd unit file should be shipped with `make install` (as I consider that to be cross-distro). Otherwise, I have kept distro-specific files outside of the release tarball because that's what we prefer in Debian (because upstream files usually try to work with many setups including old ones while downstream files only target actual releases). The init file is unlikely to change any time soon (it didn't change in 4 years), so I suppose we could keep things as they are. Note that it may be a bit outdated because it tries to work on RHEL 5, see 617233512778.

So, keep me updated when I need to update install instructions!

**Comment by hogarthj** [comment]
I did initially just take that, but with only systemd-219 in EL7 but systemd yells (but does not cause an active failure) with the directives it is not aware of.

On tossing it over in my head for a bit I figured that the CentOS/RHEL admins would prefer to avoid the noise and have a cleanly loading unit.

Incidentally ReadWritePath was ReadWriteDIrectories prior to systemd-231 so using that helps increase security in earlier Fedora releases and EL7 since we can then go ProtectSystem=full as well on those rather than just "yes" :)

I actually agree with distro stuff outside of the release tarballs.

It's a happy day to me that EL5 is finally EOL and can be totally ignored ;)

Sure - I'll ping when the packages move to stable.

**Comment by vincentbernat** [comment]
The too recent directives are commented out. I don't remember the minimal version of systemd that I am trying to be compatible with, but systemd-219 should be fine with the default config as there are directive supported by systemd-211 commented out. If there is a directive unknown to systemd-219, this is a bug on my side.

With systemd-232, we can even go to `ProtectSystem=strict`.

**Comment by hogarthj** [comment]
Huh ... I'll take another look as I'd generally rather take your upstream unit :)

Yup ... the lldpd in Fedora 26 has strict in place :)

**Comment by hogarthj** [comment]
The package is stable in all Fedora repos and as of the next EPEL push will be in the stable EPEL6 and EPEL7 repos, so the documentation can just change to yum/dnf install lldpd on centos, rhel and fedora :)

https://apps.fedoraproject.org/packages/lldpd

**Comment by vincentbernat** [comment]
❦ 21 avril 2017 03:33 -0700, James Hogarth <notifications@github.com> :

> The package is stable in all Fedora repos and as of the next EPEL push
> will be in the stable EPEL6 and EPEL7 repos, so the documentation can
> just change to yum/dnf install lldpd on centos, rhel and fedora :)
>
> https://apps.fedoraproject.org/packages/lldpd

Thanks! I'll update the documentation ASAP! Is EPEL automatically
enabled or should I tell people to also use "yum install epel-release"?
-- 
Write and test a big program in small pieces.
            - The Elements of Programming Style (Kernighan & Plauger)

**Comment by hogarthj** [comment]
On CentOS it's just a matter of yum install -y epel-release

On RHEL you need to install the release rpm and enable the optional repo
specially

https://fedoraproject.org/wiki/EPEL

**Comment by vincentbernat** [comment]
Thanks! I have updated the [instructions](http://vincentbernat.github.io/lldpd/installation.html#fedora-rhel-centos)

**Comment by hogarthj** [comment]
Great stuff, thanks :)

When it goes stable I'll close this issue and on your next release test
using your unit directly rather than modifying it :)

**Comment by boomertsfx** [comment]
Is there any way to add the bash completion stuff back into the EPEL release (maybe the RPM scripts can check to see if bash-completion is installed, and if so copy over the \usr\share\bash-completion\completions\lldpcli file)?  That was nice to have it installed by Vincent’s RPMs.

As a side note – Vincent, I’m so glad I found your tool…. It makes everything so much easier to troubleshoot, and the ability to do CDP is awesome, too!

Thanks,

Graeme

**Comment by hogarthj** [comment]
It's on my to-do list :)

Just wanted to get it packaged and out there first.

I'll do an update soon with the completion stuff in place.

**Comment by vincentbernat** [comment]
@hogarthj can this issue be closed or are you still waiting on something?

**Comment by vincentbernat** [comment]
So, it's available in Fedora and EPEL: https://apps.fedoraproject.org/koschei/package/lldpd.
